modules/core/shared/src/main/scala-3/io/circe/derivation/ConfiguredDecoder.scala

```diff
@@ -19,10 +20,22 @@ trait ConfiguredDecoder[A](using conf: Configuration) extends Decoder[A]:
 
   /** Decodes a class/object/case of a Sum type handling discriminator and strict decoding. */
   private def decodeSumElement[R](c: HCursor)(fail: DecodingFailure => R, decode: Decoder[A] => ACursor => R): R =
+
     def fromName(sumTypeName: String, cursor: ACursor): R =
-      constructorNames.indexOf(sumTypeName) match
-        case -1 => fail(DecodingFailure(s"type $name has no class/object/case named '$sumTypeName'.", cursor.history))
-        case index => decode(elemDecoders(index).asInstanceOf[Decoder[A]])(cursor)
+      def findDecoder(decoder: Decoder[?], searched: List[ConfiguredDecoder[?]]): Option[Decoder[?]] =
```

Comment:
It looks to me like the logic in here doesn't take into consideration if the `ConfiguredDecoder` is a Sum or Product decoder, and will accidentally use the decoder for a field from a Product type rather than only looking at the class names of the Sums.  

Here is 2 tests, the second is without a configured discriminator and works fine:

```scala

  {
    sealed trait GreatGrandParent
    object GreatGrandParent:
      given Eq[GreatGrandParent] = Eq.fromUniversalEquals[GreatGrandParent]

    sealed trait GrandParent extends GreatGrandParent
    case class Uncle(Child: Int) extends GrandParent
    sealed trait Parent extends GrandParent
    
    case class Child(a: Int, b: String) extends Parent

    test("Codec for hierarchy of more than 2 level with discriminator should encode and decode correctly, even if a parent's sibling has a field with the same name as a Child type") {
      given Configuration = Configuration.default.withDiscriminator("type")
      given Codec.AsObject[GreatGrandParent] = ConfiguredCodec.derived[GreatGrandParent]

      val child: GrandParent = Child(1, "a")
      val json = Encoder.AsObject[GreatGrandParent].apply(child)
      val result = Decoder[GreatGrandParent].decodeJson(json)
      assert(result === Right(child), result)
    }

    test("Codec for hierarchy of more than 2 level should encode and decode correctly, even if a parent's sibling has a field with the same name as a Child type") {
      given Configuration = Configuration.default
      given Codec.AsObject[GreatGrandParent] = ConfiguredCodec.derived[GreatGrandParent]

      val child: GrandParent = Child(1, "a")
      val json = Encoder.AsObject[GreatGrandParent].apply(child)
      val result = Decoder[GreatGrandParent].decodeJson(json)
      assert(result === Right(child), result)
    }
  }
```

Here is the result:

```
  + Codec for hierarchy of more than 1 level with discriminator should encode and decode correctly 0.007s
==> X io.circe.ConfiguredDerivesSuite.Codec for hierarchy of more than 2 level with discriminator should encode and decode correctly, even if a parent's sibling has a field with the same name as a Child type  0.047s munit.FailException: /Users/darren.gibson/Projects/circe/circe/modules/tests/shared/src/test/scala-3/io/circe/ConfiguredDerivesSuite.scala:467
466:      val result = Decoder[GreatGrandParent].decodeJson(json)
467:      assert(result === Right(child), result)
468:    }
Left(
  value = DecodingFailure at : Int
)
    at munit.FunSuite.assert(FunSuite.scala:11)
    at io.circe.ConfiguredDerivesSuite.$init$$$anonfun$27(ConfiguredDerivesSuite.scala:467)
  + Codec for hierarchy of more than 2 level should encode and decode correctly, even if a parent's sibling has a field with the same name as a Child type 0.026 
```

Reply:
Good catch, thank you @zarthross. I was trying to avoid breaking binary compatibility with those hacky implementations. 
I just found of a way to expand the internal API while preserving bin compat. 
Also added your test.